src/MongoDB.Driver.Core/Core/Authentication/AwsSignatureVersion4.cs

```diff
@@ -0,0 +1,203 @@
+﻿/* Copyright 2020–present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace MongoDB.Driver.Core.Authentication
+{
+    /// <summary>
+    /// The AWS signature version 4.
+    /// </summary>
+    public class AwsSignatureVersion4
+    {
+        /// <summary>
+        /// Gets canonical headers.
+        /// </summary>
+        /// <param name="requestHeaders">The request headers.</param>
+        /// <returns>The canonical headers.</returns>
+        public static string GetCanonicalHeaders(SortedDictionary<string, string> requestHeaders)
+        {
+            return string.Join("\n", requestHeaders.Select(x => $"{x.Key.ToLowerInvariant()}:{x.Value}"));
+        }
+
+        /// <summary>
+        /// Gets Amazon region from host.
+        /// </summary>
+        /// <param name="host">The host.</param>
+        /// <returns>The region.</returns>
+        public static string GetRegion(string host)
+        {
+            if (host == "sts.amazonaws.com")
+            {
+                return "us-east-1";
+            }
+
+            var split = host.Split('.');
+            if (split.Count() > 1)
+            {
+                return split[1];
+            }
+
+            return "us-east-1";
+        }
```

Comment:
From specification ([link](https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#conversation-5)):

> Region is not a header, but simply part of the authorization header. Region by default is ‘us-east-1’ since this is the implicit region for ‘sts.amazonaws.com’. Drivers will need to derive the region to use from the endpoint. The region is the second piece of a FQDN name. While all official AWS STS endpoints start with “sts.”, there are non-AWS hosted endpoints and test endpoints that will not follow this rule.
```
Host              | Region
sts.amazonaws.com | us-east-1
first.second      | second
first             | us-east-1
```